File: DataStructureAlgorithm/BinarySearch/BinarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

class BinarySearch{
    vector<char> arr;
    int left, mid, right, count;

public:
    BinarySearch(int size){
        for(int i=1; i<=size; ++i)
            arr.push_back(i);
        count=left=mid=0;
        right=size-1;
    }
    int find(int value){
        int tmp=right;
        while(left<=right){
            ++count;
            mid=left+(right-left)/2;
            if(arr[mid]==value){
                cout<<"ok, count: "<<count<<", ";
                count=left=0;
                right=tmp;
                return mid;
            }else if(arr[mid]<value) left=mid+1;
            else right=mid-1;
        }
        cout<<"fail, count: "<<count<<", ";
        return -1;
    }
};

int main(){
    BinarySearch a(10);
    for(int i=1; i<11; ++i){
        cout<<"value("<< i <<") index: "<<a.find(i)<<endl;
    }
    return 0;
}

File: DataStructureAlgorithm/BinarySearch/binarySearch.go
package main

import "fmt"

func main() {
	a := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	for i := 1; i <= 9; i++ {
		v, err := BinarySearch(a, i)
		if err < 0 {
			break
		} else {
			fmt.Printf("value index: %d\n", v)
		}
	}
}

func BinarySearch(arr []int, value int) (int, int) {
	var left, mid, right, count = 0, 0, len(arr) - 1, 1
	fmt.Printf("value(%d)=> ", value)
	for ; left <= right; count++ {
		mid = left + (right-left)/2
		if arr[mid] == value {
			fmt.Printf("ok, count: %d, ", count)
			return mid, 1
		} else if arr[mid] > value {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	fmt.Printf("fail, count: %d, ", count)
	return mid, -1
}
